core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -0,0 +1,610 @@
+                }
+                indexMetaDatas = new ArrayList<>(indexAliasMap.size());
+                for (ObjectObjectCursor<String, AliasMetaData> cursor : indexAliasMap) {
+                    indexMetaDatas.add(metaData.getIndices().get(cursor.key));
+                }
+            } else {
+                indexMetaDatas = Collections.singletonList(indexMetaData);
+            }
+
+            for (IndexMetaData found : indexMetaDatas) {
+                if (found.getState() == IndexMetaData.State.CLOSE) {
+                    if (failClosed) {
+                        throw new IndexClosedException(new Index(found.getIndex()));
+                    } else {
+                        if (options.forbidClosedIndices() == false) {
+                            concreteIndices.add(found.getIndex());
+                        }
+                    }
+                } else if (found.getState() == IndexMetaData.State.OPEN) {
+                    concreteIndices.add(found.getIndex());
+                } else {
+                    throw new IllegalStateException("index state [" + found.getState() + "] not supported");
+                }
+            }
+        }
+
+        if (options.allowNoIndices() == false && concreteIndices.isEmpty()) {
+            throw new IndexMissingException(new Index(Arrays.toString(indexExpressions)));
+        }
+        return concreteIndices.toArray(new String[concreteIndices.size()]);
+    }
+
+    /**
+     * Utility method that allows to resolve an index expression to its corresponding single concrete index.
+     * Callers should make sure they provide proper {@link org.elasticsearch.action.support.IndicesOptions}
+     * that require a single index as a result. The indices resolution must in fact return a single index when
+     * using this method, an {@link IllegalArgumentException} gets thrown otherwise.
+     *
+     * @param request   request containing the index or alias to be resolved to concrete index and
+     *                  the indices options to be used for the index resolution
+     * @throws IndexMissingException    if the resolved index or alias provided doesn't exist
+     * @throws IllegalArgumentException if the index resolution lead to more than one index
+     * @return the concrete index obtained as a result of the index resolution
+     */
+    public String concreteSingleIndex(ClusterState state, IndicesRequest request) throws IndexMissingException, IllegalArgumentException {
+        String index = request.indices() != null && request.indices().length > 0 ? request.indices()[0] : null;
+        return concreteSingleIndex(state, index, request.indicesOptions());
+    }
+
+    public String concreteSingleIndex(ClusterState state, DocumentRequest request) throws IndexMissingException, IllegalArgumentException {
```

Comment:
I think DocumentRequest could potentially extend IndicesRequest, then we would not need this specific method. maybe a different PR though?